RFC: DQMStore index logic rewrite

schneiml opened the pull request:
This PR attempts to complete the complete the conversion from #22157, via #22218 and #22243 et. al. It eliminates `streamId` and `enableMultiThread` from the DQMStore.

This needs to be merged/cherry-picked with #22319 and #22364, it adds another change on top: re-doing the logic that queries MEs out of the set in the DQMStore.

It needs to be discussed if the handling of the runNumber is actually what we want.

This is likely to break multi-run harvesting and online (needs testing).

Commit:
Disable multi thread for ALCAHARVEST.

schneiml committed:
Merged refs/pull/22285/head from repository cms-sw with cms-merge-topic

schneiml committed:
Switch to EDProducer.

Also re-introduce the .cc file, implementations in the header are to likely to confuse the build system.

schneiml committed:
Introduce more DQMTokens.

For now, step1 and Harvesting produce the same token, which is fine since we don't need to handle Sources and Harvesters in the same process yet. That way, the output modules will for sure wait for everything.

There is a new type of token for the endRun, to make sure Lumi and Run transistions get ordered correctly.

schneiml committed:
Convert EDMtoMEConverter to a producer, similar to DQMEDHarvester.

Also protect the DQMStore using usesResource, since all of this is
really dangerous.

schneiml committed:
Some formal changes to MEtoEDMConverter.

Consume tokens, protect the DQMStore and watch all the transitions.

schneiml committed:
Another DQMEDAnalyzer config change

schneiml committed:
HACK: make MEtoEDMConverter not delete things, and force-off that enableMultiThread switch.

Does not fix things, however.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22307/3508

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for master.

It involves the following packages:

Configuration/PyReleaseValidation
DQMServices/Components
DQMServices/Core
DataFormats/Histograms
Validation/RecoTau


@smuzaffar, @prebello, @Dr15Jones, @vazzolini, @kmaeshima, @kpedro88, @fabozzi, @cmsbuild, @jfernan2, @GurpreetSinghChahal, @vanbesien, @dmitrijus can you please review it and eventually sign? Thanks.
@ghellwig, @barvic, @makortel, @felicepantaleo, @rovere, @Martin-Grunewald, @ebrondol this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented on DQMServices/Components/plugins/MEtoEDMConverter.cc:
```diff
@@ -81,6 +81,8 @@ MEtoEDMConverter::MEtoEDMConverter(const edm::ParameterSet & iPSet) :
   produces<MEtoEDM<TString>, edm::Transition::EndLuminosityBlock>(sName);
 
   consumesMany<DQMToken>();
+  consumesMany<DQMRunToken>();
```
These need to be 
```
consumesMany<DQMToken, edm::InLumi>();
consumesMany<DQMRunToken, edm::InRun>();
```

fwyzard replied:
in fact, I guess we could just use `DQMToken` for both Run and Lumi MEs, and let the framework differentiate based on the `InLumi` and `InRun` flags.

schneiml replied:
ok, let me try. _Maybe_ this fixes our problmes...

schneiml replied:
obviously that would be too easy. Fixed, but #22281 continues...

Dr15Jones commented on DataFormats/Histograms/interface/DQMToken.h:
```diff
@@ -25,5 +25,12 @@ class DQMToken
       DQMToken() {}
 };
 
+class DQMRunToken
```
There really is not a need for a seperate type just for Run.

schneiml replied:
@Dr15Jones I thought so. But having two `produce` calls with the same Tokens gave me an error, and this was the quick fix.

fwyzard commented on DQMServices/Core/src/DQMEDAnalyzer.cc:
```diff
@@ -0,0 +1,53 @@
+    run.moduleCallingContext()->moduleDescription()->id());
+}
+
+void DQMEDAnalyzer::endRun(edm::Run const& run, edm::EventSetup const& setup) 
```
I would suggest to keep most methods - especially the empty ones - in the .h file.

schneiml replied:
I in general prefer that -- but if it gives me linker errors, I happily put them into a `.cc` file. My theory is that the build system gets more easily confused with implementation in headers (and honestly, I have no idea how this is supposed to work on the linker level, so putting back the `.cc` file was the obvious option).

fwyzard replied:
No, the build system does not get "confused".

If you put the implementation of a method inside the class body, e.g.
```c++
class Class {
public:
    void some_method() { ... }
};
```
it is automatically `inline`, and should not create linker problems.

If you put non-inline out-of-body implementation, like
```c++
class Class {
public:
    void some_method();
};

void Class::some_method() { ... }
```
in the .h file, then you do get one version in each .cc file that includes it, and the linker will rightfully compile.

One other thing to pay attention to is if you have virtual methods,  you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

I would be curious what problem you ran into here, and if we can find a less drastic solution than moving all methods' implementations to the .cc file.

schneiml replied:
> One other thing to pay attention to is if you have virtual methods, you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

This one. So my gut feeling of putting things into a .cc file was justifed.

And yes, this inline story works fine as long as the methods are not virtual. Then it gets interesting.

schneiml commented on DQMServices/Core/src/DQMEDAnalyzer.cc:
```diff
@@ -0,0 +1,53 @@
+
+DQMEDAnalyzer::DQMEDAnalyzer() {
+  produces<DQMToken,edm::Transition::EndLuminosityBlock>();
+  produces<DQMRunToken,edm::Transition::EndRun>();
```
@Dr15Jones when I remove the `Run` in this line, I got an framework error, saying I cannot have more than on produce call for the same product. I guess I have to specify both transitions in one call?

Dr15Jones replied:
Hmm. Our framework test works: 
https://github.com/cms-sw/cmssw/blob/ede02fc7b4740a09bb6e578d9ec71129d8c19b19/FWCore/Integration/test/ThingProducer.cc#L14

However, that does use unique product instance names, which was not meant to be required.

fwyzard commented on DQMServices/Components/plugins/MEtoEDMConverter.cc:
```diff
@@ -92,7 +94,7 @@ void
 {
   // Determine if we are running multithreading asking to the DQMStore. Not to be moved in the ctor
-  enableMultiThread_ = dbe->enableMultiThread_;
+  //enableMultiThread_ = dbe->enableMultiThread_;
```
just delete the commented-out lines (and drop the empty methods)

schneiml replied:
This commit says `HACK`  in the message for a good reason. After #22218, `enableMultiThread` has exactly no reason for existence, but I'll delay removing it everywhere to after #22281 is resolved.

fwyzard commented on DQMServices/Components/plugins/MEtoEDMConverter.h:
```diff
@@ -51,8 +51,10 @@
                                                      edm::EndLuminosityBlockProducer,
-                                                     edm::EndRunProducer>
+                                                     edm::EndRunProducer,
+                                                     edm::one::SharedResources>
```
what is the purpose of the `SharedResources` ?

Dr15Jones replied:
I also do not understand its usage.

schneiml replied:
It gives the `usesResource` call, which is a quick way to get a global lock for legacy stuff. Exactly what legacy DQM needs.

Though ideally, the final version of this PR will simply make that all thread save and we can get rid of `usesResource("DQMStore")` everywhere.

fwyzard commented:
I don't understand one thing: the DQMEDProducer now produces the token, and the METoEDM consumes it.
But what about the DQMEDHarvesters ?

fwyzard commented:
Naively I would say that the harvesters should condume only the token(s) of the module(s) that produce the plots they want to harvest; and produce new tokens of their own.

Dr15Jones commented:
I agree with @fwyzard about the behavior of harvesters

fwyzard commented:
@dr15jones, how do I make the FastTimerService produce a DQMToken ?
or, should i just call the DQMStore methods at the right moments and let the framework ignore the dependencies ?

schneiml commented:
@Dr15Jones, @fwyzard -- we agree that we need two types of tokens, one for the "Source" plugins, and one for harvesting. But for now the priority was to get #22281 fixed. This PR is not really done yet, but I wanted to publish the ton of fixes towards fixing ALCA, since likely we'll need others help.

Dr15Jones commented:
>@Dr15Jones, how do I make the FastTimerService produce a DQMToken ?
or, should i just call the DQMStore methods at the right moments and let the framework ignore the dependencies ?

Services are not allowed to `produce` data. Therefore either
-  talk to the DQMStore directly or
-  have another EDProducer which can get called at the appropriate time?

schneiml commented:
> Naively I would say that the harvesters should condume only the token(s) of the module(s) that produce the plots they want to harvest; and produce new tokens of their own.

Ah indeed, I mis-read you there. This is where we want to eventually end up, but this will not be possible without subsystem help. Which means the timescale is ~years. 

That would also allow us to specify proper dependencies between harvesters, which currently... they might be there, and scheduled manually, but maybe broken after the harvesting producer migration... We don't really know. (I think there are close to none of these dependencies atm, since it is much more obvious to just do all harvesting in one plugin.)

schneiml committed:
Improve consumes, thanks @Dr15Jones

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22307/3512

cmsbuild commented:
Pull request #22307 was updated. @smuzaffar, @prebello, @Dr15Jones, @vazzolini, @kmaeshima, @kpedro88, @fabozzi, @cmsbuild, @jfernan2, @GurpreetSinghChahal, @vanbesien, @dmitrijus can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

schneiml commented:
Btw @Dr15Jones : We also need to get rid of the underscores in the plugin labels of the soon-to-be EDProducers? That would be another big PR (~60 files in subsystem code).

But it seems to mostly work for now, even with underscores...

cmsbuild commented:
-1

Tested at: 3b738e1a138e3707b383b57ab7a4a7434ceb38f6


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/d7247097a3cf98164b8b2c1a38ef0bd92d943195
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22307/26245/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22307/26245/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22307/26245/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17.log

</pre>20434.0 step3
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/d7247097a3cf98164b8b2c1a38ef0bd92d943195
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22307/26245/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22307/26245/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

schneiml committed:
Never set enableMultiThread.

schneiml committed:
Remove enableMultiThread from the DQMStore.

This involved fixing the save() logic to understand the new meaning of moduleId.

schneiml committed:
Remove enableMultiThread from DQMStore

schneiml committed:
Fix some python errors

fwyzard commented:
> Btw @Dr15Jones : We also need to get rid of the underscores in the plugin labels of the soon-to-be EDProducers? That would be another big PR (~60 files in subsystem code).

I would guess yes, as an underscore in the module label will "confuse" the branch name of the product.

It may work fine as long as you use only `consumesMany`, and do not reference the labels directly.
It may still create problems with the process name, though.

schneiml committed:
Force-0 streamId.

They serve no purpose any longer, only cause confusion, and should be eradicated.

schneiml committed:
Add new logic to getAllContents()

schneiml committed:
Completely eradicate the streamId

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22307/3543

cmsbuild commented:
Pull request #22307 was updated. @smuzaffar, @prebello, @Dr15Jones, @vazzolini, @kmaeshima, @kpedro88, @fabozzi, @cmsbuild, @jfernan2, @GurpreetSinghChahal, @vanbesien, @dmitrijus can you please check and sign again.

schneiml commented:
This is still WIP and probably does not even compile yet.

schneiml committed:
Replace custom logic in save() with getAllContents()

... and fix the logic of it.

fwyzard commented:
@schneiml , if you agree that this is superseded by #22319, could you close it ?

schneiml commented:
@fwyzard they are rather diverged by now, because I was doing things that are no longer covered by the title. I'll close it, and make a new PR once theses things are done.

schneiml committed:
Fix more cases in the getAllContents logic.

There might be more missing, per-lumi things, online, and multi-run harvesting. But the usual sequences run fine, now multi-threaded.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22307/3582

cmsbuild commented:
Pull request #22307 was updated. @smuzaffar, @prebello, @Dr15Jones, @vazzolini, @kmaeshima, @kpedro88, @fabozzi, @cmsbuild, @jfernan2, @GurpreetSinghChahal, @vanbesien, @dmitrijus can you please check and sign again.
